### ISSUE TITLE
Undo-tree breaks with polymode

### ISSUE THREAD

**Issue by jody-frankowski** [polymode/polymode#214]
Hi,

There seems to be a conflict with undo-tree and polymode.
In a nutshell: `u` and `C-r` (which are evil's way of undoing and redoing) need to be done in the same "segments".

init.el:
```elisp
(require 'package)
(package-initialize)

(setq package-archives '(("org"   . "http://orgmode.org/elpa/")
                            ("gnu"   . "http://elpa.gnu.org/packages/")
                            ("melpa" . "https://melpa.org/packages/")))
(package-refresh-contents)
(unless (package-installed-p 'use-package)
    (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(use-package evil
  :config
  (evil-mode 1))
;; evil already pulls undo-tree, but i'm being explicit here.
(use-package undo-tree
  :init
  (unless (file-exists-p "/tmp/undo")
    (make-directory "/tmp/undo"))
  :config
  (setq undo-tree-auto-save-history t
        undo-tree-enable-undo-in-region nil
        undo-tree-history-directory-alist '(("." . "/tmp/undo"))))

(use-package poly-org)
```

test.org:
```org-mode
AAA

#+BEGIN_SRC emacs-lisp
(use-package poly-org)
(use-package poly-markdown)
#+END_SRC
```

Execute the following:

```sh
emacs -nw -q --load init.el test.org
```

Then type the following keybindings (without the spaces) for two working test cases:

```
dd u C-r
```
```
dd jj u jjj C-r
```

Here is the non working one:

```
dd jj u C-r
```

I have tested the same things without polymode, and all the tests pass (even though undo-tree is notorious for breaking unexpectedly)
I guess the same can be done without undo-tree, it's just that undo-tree is a hard dependency of evil-mode for now (https://github.com/emacs-evil/evil/issues/1074)

**Comment by vspinu** [comment]
Well, the default undo in emacs is already complicated enough. I am happy that I made it work across different buffers. From what I read undo-tree is complex and buggy and that that makes me not to want spending time on it.  But I expect that fixing this issue amounts to adding the data structure that undo-tree uses internally to the `polymode-move-these-vars-from-old-buffer`. The built-in `buffer-undo-list` is part of that list already. Could you identify this data structure?

**Comment by jody-frankowski** [comment]
I think that would be `buffer-undo-tree`. I tried with this variable and it seems to work fine now :)

**Comment by vspinu** [comment]
Perfect. Thanks! I have added it to the default list.

**Comment by jody-frankowski** [comment]
Great! Thank you very much!

**Comment by kurnevsky** [comment]
Also additional useful hack for undo tree - initialize `buffer-undo-tree` in mode hook, like:
```elisp
(add-hook 'markdown-mode-hook (lambda () (when (null buffer-undo-tree) (setq buffer-undo-tree (make-undo-tree)))))
```

btw, is there hook for any polymode?

**Comment by vspinu** [comment]
Why is this needed? 

> btw, is there hook for any polymode?

Yes. Every polymode has a poly-xyz-mode-hook which runs in every polymode buffer and there are more specialized hooks which run on specific modes or buffer switch. What do you need more concretely?

**Comment by kurnevsky** [comment]
@vspinu I'm not sure what exactly happens but until first undo `buffer-undo-tree` is nil. `undo-tree-undo` checks if it's nil and if so initializes it with `(setq buffer-undo-tree (make-undo-tree))`. And in this case first call to `undo-tree-undo` inside an indirect buffer will lead to lost undo history.  So perhaps copying doesn't work correct or not enough in this case. That's why I want to initialize it explicitly in each buffer where polymode is active.

**Comment by vspinu** [comment]
As far as I can see this initialization should not be needed during the interactive use. Nor the explicit initialization as you propose should solve the lost history problem because you initialize it to an empty tree anyhow. 

I have just made a small change to the core. Could you please check if the problem persists?

**Comment by kurnevsky** [comment]
@vspinu it still persists, and this explicit initialization somehow fixes it. It also works without initialization but only after first undo being made when it's initialized by undo-tree. I can make an instruction to reproduce it if you want to investigate - it's 100% reproducible.

**Comment by vspinu** [comment]
> I can make an instruction to reproduce 

Would be great. Please open a new issue. Thanks.
